sycl/include/CL/sycl/detail/image_accessor_util.hpp

```diff
@@ -300,92 +297,84 @@ void convertReadData(vec<ChannelType, 4> PixelData,
     throw cl::sycl::invalid_parameter_error(
         "Datatype of read data - cl_int4 is incompatible with the "
         "image_channel_type of the image.");
-    break;
   }
 }
 
 template <typename ChannelType>
-void convertReadData(vec<ChannelType, 4> PixelData,
-                     image_channel_type ImageChannelType,
+void convertReadData(const vec<ChannelType, 4> PixelData,
+                     const image_channel_type ImageChannelType,
                      vec<cl_float, 4> &RetData) {
 
-  vec<cl_float, 4> Temp;
-
   switch (ImageChannelType) {
   case image_channel_type::snorm_int8:
     //  max(-1.0f, (float)c / 127.0f)
-    RetData.x() = static_cast<cl_float>(PixelData.x()) / 127.0f;
-    RetData.y() = static_cast<cl_float>(PixelData.y()) / 127.0f;
-    RetData.z() = static_cast<cl_float>(PixelData.z()) / 127.0f;
-    RetData.w() = static_cast<cl_float>(PixelData.w()) / 127.0f;
+    RetData = (PixelData.template convert<cl_float>()) / 127.0f;
     RetData = cl::sycl::fmax(RetData, -1);
     break;
   case image_channel_type::snorm_int16:
     // max(-1.0f, (float)c / 32767.0f)
-    RetData.x() = static_cast<cl_float>(PixelData.x()) / 32767.0f;
-    RetData.y() = static_cast<cl_float>(PixelData.y()) / 32767.0f;
-    RetData.z() = static_cast<cl_float>(PixelData.z()) / 32767.0f;
-    RetData.w() = static_cast<cl_float>(PixelData.w()) / 32767.0f;
+    RetData = (PixelData.template convert<cl_float>()) / 32767.0f;
     RetData = cl::sycl::fmax(RetData, -1);
     break;
   case image_channel_type::unorm_int8:
     // (float)c / 255.0f
-    RetData.x() = static_cast<cl_float>(PixelData.x()) / 255.0f;
-    RetData.y() = static_cast<cl_float>(PixelData.y()) / 255.0f;
-    RetData.z() = static_cast<cl_float>(PixelData.z()) / 255.0f;
-    RetData.w() = static_cast<cl_float>(PixelData.w()) / 255.0f;
+    RetData = (PixelData.template convert<cl_float>()) / 255.0f;
     break;
   case image_channel_type::unorm_int16:
     // (float)c / 65535.0f
-    RetData.x() = static_cast<cl_float>(PixelData.x()) / 65535.0f;
-    RetData.y() = static_cast<cl_float>(PixelData.y()) / 65535.0f;
-    RetData.z() = static_cast<cl_float>(PixelData.z()) / 65535.0f;
-    RetData.w() = static_cast<cl_float>(PixelData.w()) / 65535.0f;
+    RetData = (PixelData.template convert<cl_float>()) / 65535.0f;
     break;
-  case image_channel_type::unorm_short_565:
+  case image_channel_type::unorm_short_565: {
     // TODO: Missing information in OpenCL spec. check if the below code is
     // correct after the spec is updated.
     // Assuming: (float)c / 31.0f; c represents the 5-bit integer.
     //           (float)c / 63.0f; c represents the 6-bit integer.
 
     // Extracting each 5/6-bit channel data.
     // PixelData.x will be of type cl_ushort.
+
+    vec<cl_float, 4> Temp;
```

Comment:
I would prefer to see this code like this, perhaps with nicer names.

```
cl_int4 Shifter{11, 5, 0, 0};
cl_ushort4 ForBitAnd = cl_ushort4{0xF800, 0x07E0, 0x001F, 1};
cl_ushort4 ConvertedPixelData = PixelData.template convert<cl_ushort>();
cl_ushort4 IntermediateResult = ConvertedPixelData & ForBitAnd;
cl_float4 IntermediateResult2 = IntermediateResult >> Shift;
cl_float4 StrangeDivider = cl_float4{31.0f, 63.0f, 31.0f, 1.0f};
RetData = IntermediateResult2 / StrangeDivider;
```
In this case, i guess, if the host compiler supports vector operations, it can better vectorize this code.
and no casts :)

Reply:
done with no casts :D.